File: haskell/glue/src/Glue/Lib/Math/Arithmetic.hs
module Glue.Lib.Math.Arithmetic where

import Glue.Eval (Eval)
import Glue.IR (IR (..))
import Glue.Lib.Math.Arithmetic.Add qualified as Add
import Glue.Lib.Math.Arithmetic.Div qualified as Div
import Glue.Lib.Math.Arithmetic.Mod qualified as Mod
import Glue.Lib.Math.Arithmetic.Mul qualified as Mul
import Glue.Lib.Math.Arithmetic.Sub qualified as Sub
import Glue.Module (ModuleInfo, nativeModule)

arithmetic :: ModuleInfo Eval
arithmetic =
    nativeModule
        "ffi.math.arithmetic"
        []

File: dart/glue/lib/src/lib/math/arithmetic.dart
import 'package:glue/src/lib/math/arithmetic/add.dart';
import 'package:glue/src/lib/math/arithmetic/div.dart';
import 'package:glue/src/lib/math/arithmetic/mod.dart';
import 'package:glue/src/lib/math/arithmetic/mul.dart';
import 'package:glue/src/lib/math/arithmetic/sub.dart';
import 'package:glue/src/module.dart';

/// Arithmetic module - basic mathematical operations
/// Mirrors Haskell Glue.Lib.Math.Arithmetic exactly

/// The arithmetic module containing all basic math operations
/// Mirrors Haskell Glue.Lib.Math.Arithmetic.arithmetic exactly
final ModuleInfo arithmeticModule = nativeModule('ffi.math.arithmetic', [
  // Addition
  ('+', add),
  ('add', add),

  // Subtraction
  ('-', sub),
  ('sub', sub),

  // Multiplication
  ('*', mul),
  ('mul', mul),

  // Division
  ('/', div),
  ('div', div),

  // Modulo
  ('%', mod),
  ('mod', mod),
]);
